safekeeper/src/pull_timeline.rs

```diff
@@ -158,13 +166,51 @@ impl WalResidentTimeline {
     async fn start_snapshot<W: AsyncWrite + Unpin + Send>(
         &self,
         ar: &mut tokio_tar::Builder<W>,
+        source: NodeId,
+        destination: NodeId,
     ) -> Result<SnapshotContext> {
         let mut shared_state = self.write_shared_state().await;
         let wal_seg_size = shared_state.get_wal_seg_size();
 
         let cf_path = self.get_timeline_dir().join(CONTROL_FILE_NAME);
-        let mut cf = File::open(cf_path).await?;
-        ar.append_file(CONTROL_FILE_NAME, &mut cf).await?;
+
+        let mut control_store = control_file::FileStorage::load_control_file(&cf_path)?;
+        let replace = control_store
+            .partial_backup
+            .replace_uploaded_segment(source, destination);
+
+        if let Some(replace) = replace {
+            // The deserialized control file has an uploaded partial. We upload a copy
+            // of it to object storage for the destination safekeeper and send an updated
+            // control file in the snapshot.
+            tracing::info!(
+                "Replacing uploaded partial segment in in-mem control file: {replace:?}"
+            );
+
+            let remote_timeline_path = wal_backup::remote_timeline_path(&self.tli.ttid)?;
+            wal_backup::copy_partial_segment(
+                &replace.previous.remote_path(&remote_timeline_path),
+                &replace.current.remote_path(&remote_timeline_path),
+            )
+            .await?;
+
+            scopeguard::defer! {
+                if let Err(err) = control_file::FileStorage::remove_temp(self.get_timeline_dir()) {
+                    tracing::warn!("Failed to remove temp control file: {err}");
+                }
+            }
+
+            let mut control_temp =
+                control_file::FileStorage::write_temp(&control_store, self.get_timeline_dir())
```

Comment:
What do you think about not using `write_temp` to create a temp file on disk and instead try to write bytes from `write_to_buf` directly to archive?

https://docs.rs/tokio-tar/latest/tokio_tar/struct.Builder.html#method.append_data

I looked briefly at tokio-tar and I see that API for this is not great, but writing a temp file to disk feels too hacky.

Reply:
> What do you think about not using write_temp to create a temp file on disk and instead try to write bytes from write_to_buf directly to archive?

Yeah, that's what I tried initially, but couldn't figure out how `Header` should look like. I can look into it again

Reply:
They have this snippet in the doc:
```
let mut header = Header::new_gnu();
header.set_size(4);
header.set_cksum();

let mut data: &[u8] = &[1, 2, 3, 4];

let mut ar = Builder::new(Vec::new());
ar.append_data(&mut header, "really/long/path/to/foo", data).await?;
let data = ar.into_inner().await?;
```

Reply:
I wasn't sure about file perms, but I think it's fine. Done in https://github.com/neondatabase/neon/pull/8628/commits/fb88c56277d47a96ced3a88c4242e1aa3507765e